### PR TITLE
fix docstring, add getting available port number

### PR DESCRIPTION
**Pull request by KazuCocoa**

### CODE REVIEW

File: test/functional/ios/webdriver_tests.py [hunk 1]
```diff
@@ -24,20 +24,23 @@ class WebDriverTests(unittest.TestCase):
...
-        if self.driver1:
-            self.driver1.quit()
+        desired_caps['wdaLocalPort'] = port
+
+        driver2 = None
+        try:
+            driver2 = webdriver.Remote('http://localhost:4723/wd/hub', desired_caps)
+            self.assertEqual(2, len(self.driver.all_sessions))
+        finally:
+            if driver2:
```
Comment (mykola-mokhnach) [inline test/functional/ios/webdriver_tests.py]: is not None

File: test/functional/ios/helper/desired_capabilities.py [hunk 2]
```diff
@@ -63,10 +63,23 @@ def wda_port():
...
     return 8100
 
+
+def get_available_port():
+    """Returns available local port number.
+    """
+    import socket
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+    for port in range(8102, 8200):
```
Comment (mykola-mokhnach) [inline test/functional/ios/helper/desired_capabilities.py]: magic numbers?

File: test/functional/ios/webdriver_tests.py [hunk 3]
```diff
@@ -24,20 +24,23 @@ class WebDriverTests(unittest.TestCase):
...
         self.driver = webdriver.Remote('http://localhost:4723/wd/hub', desired_caps)
-        desired_caps['deviceName'] = 'iPhone Xs Max'
-        desired_caps['wdaLocalPort'] = '8102'
 
     def tearDown(self):
         self.driver.quit()
 
-    def testAllSessions(self):
+    def test_all_sessions(self):
+        port = desired_capabilities.get_available_port()
```
[the remark this answers is not in the document]
Comment (KazuCocoa) [reply test/functional/ios/webdriver_tests.py]: yes, let me update them to the proper place later.
Current this test is not stable (I rebuild once to make the CI green)

File: test/functional/test_helper.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+def get_available_port(in_range):
+    """Returns available local port number.
+    """
+    range(8102, 8200)
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: what this operator is for?

File: test/functional/test_helper.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    if not isinstance(in_range, range):
+        raise ValueError('{} should be range'.format(in_range))
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: should be of type range

File: test/functional/test_helper.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+def get_available_port(in_range):
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: port_range

File: test/functional/test_helper.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    """Returns available local port number.
+    """
+    range(8102, 8200)
+    if not isinstance(in_range, range):
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: in python3 range creates a generator by default. I would rather provide just start and end port numbers with default valuesof 8102 and 8200 as method arguments

File: test/functional/test_helper.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+        raise ValueError('{} should be range'.format(in_range))
+
+    import socket
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+    for port in in_range:
+        if sock.connect_ex(('localhost', port)) != 0:
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]:
```
try:
    if sock.connect_ex(('localhost', port)) != 0:
       return port
finally:
    sock.close()
```

File: test/functional/test_helper.py [hunk 9]
```diff
@@ -1,8 +1,8 @@
 def get_available_port_range(from_range, to_range):
     """Returns available local port number.
     """
-    if isinstance(from_range, int) or isinstance(to_range, int):
-        raise ValueError('from_range and to_range should be type of int')
+    if not isinstance(from_range, int) or not isinstance(to_range, int):
+        raise ValueError('"{}" and "{}" should be type of int')
```
Comment (ki4070ma) [inline test/functional/test_helper.py]: Just question, does it work?

```
ValueError('"{}" and "{}" should be type of int')
```

In my understanding, at least this works.

```
ValueError("{} and {} should be type of int".format(from_range, to_range))
```

Comment (KazuCocoa) [reply test/functional/test_helper.py]: Yes, I've wrapped up in 917b34f

File: test/functional/test_helper.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+def get_available_port_range(from_range, to_range):
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: ...port_from_range

**Comment by KazuCocoa** [comment]
remove wip

File: test/functional/test_helper.py (hunk 10 above)
Comment (mykola-mokhnach) [reply test/functional/test_helper.py]: from_port, to_port

File: test/functional/test_helper.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    """Returns available local port number.
+    """
+    if not isinstance(from_range, int) or not isinstance(to_range, int):
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: explicit types verifications is not very pythonic

simply cast values to in in `range` call and let it throw if it is not possible

File: test/functional/test_helper.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+        raise ValueError('"{}" and "{}" should be type of int'.format(from_range, to_range))
+
+    import socket
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+    for port in range(from_range, to_range):
+        try:
+            if sock.connect_ex(('localhost', port)) != 0:
+                return port
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: this function is going to return None if no port has been matched

File: test/functional/ios/webdriver_tests.py [hunk 13]
```diff
@@ -24,20 +27,31 @@ class WebDriverTests(unittest.TestCase):
...
         desired_caps = desired_capabilities.get_desired_capabilities('UICatalog.app.zip')
         desired_caps['deviceName'] = 'iPhone Xs Max'
-        desired_caps['wdaLocalPort'] = '8102'
-        self.driver1 = webdriver.Remote('http://localhost:4723/wd/hub', desired_caps)
-        self.assertEqual(2, len(self.driver.all_sessions))
-        if self.driver1:
-            self.driver1.quit()
+        desired_caps['wdaLocalPort'] = port
+
+        class get_all_sessions(object):
```
Comment (ki4070ma) [inline test/functional/ios/webdriver_tests.py]: This class name doesn't mean actual behavior and it's a little tricky.
I'd rather have your original code style from maintenance point of view if it works.

```
WebDriverWait(driver2, 10).until(len(driver.all_sessions) == 2)

or

TIMEOUT = 10
WebDriverWait(driver2, TIMEOUT).until(len(driver.all_sessions) == 2)
```

Comment (KazuCocoa) [reply test/functional/ios/webdriver_tests.py]: `len(driver.all_sessions) == 2` did not work in `WebDriverWait`, unfortunately

File: test/functional/test_helper.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
+class NoAvailablePortError(Exception):
+    pass
+
+
+def get_available_from_port_range(from_port, to_port):
+    """Returns available local port number.
+    """
+    r = range(from_port, to_port)
+
+    import socket
```
Comment (mykola-mokhnach) [inline test/functional/test_helper.py]: why we don't import socket at the start of the module?

File: test/functional/ios/webdriver_tests.py [hunk 15]
```diff
@@ -24,20 +27,31 @@ class WebDriverTests(unittest.TestCase):
...
         desired_caps = desired_capabilities.get_desired_capabilities('UICatalog.app.zip')
         desired_caps['deviceName'] = 'iPhone Xs Max'
-        desired_caps['wdaLocalPort'] = '8102'
-        self.driver1 = webdriver.Remote('http://localhost:4723/wd/hub', desired_caps)
-        self.assertEqual(2, len(self.driver.all_sessions))
-        if self.driver1:
-            self.driver1.quit()
+        desired_caps['wdaLocalPort'] = port
+
+        class GetAllSessions(object):
```
Comment (mykola-mokhnach) [inline test/functional/ios/webdriver_tests.py]: usually class names are nouns. How about SessionCountIsTwoCondition ?

File: test/functional/test_helper.py (hunk 14 above)
Comment (KazuCocoa) [reply test/functional/test_helper.py]: Just copied and pasted, but can move it to outside. Will do with renaming the class name

File: test/functional/ios/webdriver_tests.py [hunk 16]
```diff
@@ -24,20 +27,31 @@ class WebDriverTests(unittest.TestCase):
...
         desired_caps = desired_capabilities.get_desired_capabilities('UICatalog.app.zip')
         desired_caps['deviceName'] = 'iPhone Xs Max'
-        desired_caps['wdaLocalPort'] = '8102'
-        self.driver1 = webdriver.Remote('http://localhost:4723/wd/hub', desired_caps)
-        self.assertEqual(2, len(self.driver.all_sessions))
-        if self.driver1:
-            self.driver1.quit()
+        desired_caps['wdaLocalPort'] = port
+
+        class session_counts_is_two(object):
```
Comment (KazuCocoa) [inline test/functional/ios/webdriver_tests.py]: followed snake case to follow https://selenium-python.readthedocs.io/waits.html 's name style

**Pull request merged**